### PR TITLE
Convert Markdown files to ReST

### PR DESCRIPTION
**Pull request by rouille**
### Purpose
Convert Markdown files to ReST files. Also, the installation guide has been updated/improved and references to the GSoC has been removed in the list of mentored projects.

### Where to look
* move content of ***introduction***, ***installation*** and ***communication*** files into ***index*** file
* update ***.makeignore*** file
* ***code_of_conduct*** file has been moved to a new **communication** folder
* ***idea_list*** file has been renamed ***project_ideas*** and moved to the **communication** folder. Mention to GSoC has been removed and the document is now titled `Mentored Project Ideas`
* installation guide has been improved, e.g., remove mention to `BASE_PROFILE_DIR`since profiles are now fetched from blob storage, etc.
* refactor code snippets in the ***coding_guide*** file to use snake_case naming and Python examples (there was some JavaScript examples)
* fix some broken links

### Time estimate
15 min

### CODE REVIEW

File: source/dev/coding_guide.rst [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        if (time.hour >= 0 and time.hour < 6):
+            return "Night"
+        if (time.hour >= 6 and time.hour < 12):
+            return "Morning"
+        if (time.hour >= 12 and time.hour < 18):
+            return "Afternoon"
+        return "Evening"
+
+This method looks simple, but it is very difficult to write a proper state-based unit
+test for it. ``DateTime.Now`` is, essentially, a hidden input that we can't control.
```
Comment (jon-hagg) [inline source/dev/coding_guide.rst:97]: One more snake case - `DateTime.Now`

File: source/dev/coding_guide.rst [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+because, in addition to their primary duties, they control their own initialization and
+lifecycle.
+
+Singletons can easily make unit tests order-dependent because they carry state around
+for the lifetime of the whole application or unit test suite. Have a look at the
+following example:
+
+.. code-block:: python
+
+    def get_user(user_id):
```
Comment (jon-hagg) [inline source/dev/coding_guide.rst:362]: I think the original version of this is almost right, except for the casing, and would change the `contains_key` to be `user_id in user_cache.instance`

File: source/dev/coding_guide.rst (hunk 1 above)
Comment (rouille) [reply source/dev/coding_guide.rst]: done

File: source/dev/coding_guide.rst (hunk 2 above)
Comment (rouille) [reply source/dev/coding_guide.rst:362]: done

**Review by jon-hagg — APPROVED** [review]

**Pull request merged**